aspnetcore/fundamentals/dependency-injection/samples/3.x/DIsample2/DIsample2/Services/Service1.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace DIsample2.Services
+{
+    public class Service1 : IDisposable
+    {
+        private bool disposedValue;
+
+        protected virtual void Dispose(bool disposing)
+        {
+            Console.WriteLine("Service1.Dispose");
```

Comment:
@serpent5  take a look at the section in the doc [Disposal of services](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.1&branch=pr-en-us-19271#disposal-of-services)

I've implemented this in aspnetcore/fundamentals/dependency-injection/samples/3.x/DIsample2 but  `Dispose` is never called.

For the singleton case, who cares if   `Dispose` is never called? At app shutdown, all the memory allocated by the process is released.

Reply:
> I've implemented this in aspnetcore/fundamentals/dependency-injection/samples/3.x/DIsample2 but `Dispose` is never called.

`Dispose` isn't called because no instances of `Service1` are created in that sample. You need to take a dependency on `Service1` for the DI container to both create and dispose of `Service1`. e.g.:

```csharp
public IndexModel(ILogger<IndexModel> logger, Service1 service1)
{
    ...
}
````